### PR TITLE
Add support for Arch/Artix Linux

### PR DESCRIPTION
**Pull request by ne9z**
Articles are directly viewable at [here](https://github.com/ne9z/openzfs-docs/tree/arch-support/docs/Getting%20Started/Arch%20Linux).

Revised layout is the following:
```
bpool/BOOT/default        on        /boot
bpool/BOOT/bootenv        on        /boot        # alternative boot environment
rpool/ROOT/default        on        /
rpool/ROOT/bootenv        on        /            # alternative boot environment
rpool/DATA/default/{home,var/log,...}            # persistent data
```

This layout enables the destruction of the initial boot environment `default`. With the previous layout, where persistent data is nested under `rpool/ROOT/default`, this is not possible.

Another addition is appending pool names with a random UUID. This will prevent naming conflict when importing pools on another root on ZFS system.

Boot environment management is performed with [this AUR package](https://aur.archlinux.org/packages/bieaz/). It will also add relevent GRUB menu entries for easy recovery.

A pacman hook for creating boot environments upon each pacman transaction is also [available](https://aur.archlinux.org/packages/rozb3-pac/).

@rlaager  @gmelikov

Signed-off-by: Maurice Zhou <ja@apvc.uk>

### CODE REVIEW

**Comment by ne9z** [comment]
Just completed a testing installation and corrected some mistakes. Should be mostly bug-free now.

Some screenshots of the boot environment management utility:
View and select a boot environment in GRUB menu:
GRUB_1
GRUB_2

Show info and list boot environments:
```
[root@localhost ~]# bieaz info
ZFS property for label:    org:bieaz
Current boot environment:  default
Root dataset container:    /ROOT/
Root dataset full path:    rpool_n9gyiw/ROOT/default
Boot dataset container:    /BOOT/
Boot dataset full path:    bpool_n9gyiw/BOOT/default
Root pool name:            rpool_n9gyiw
Boot pool name:            bpool_n9gyiw
[root@localhost ~]# bieaz list
NAME        MOUNTED   USED  CREATION               ORG:BIEAZ
pac-0znryh  no        600K  Wed Jan 13 21:06 2021  post_pacman_-S_grub
pac-qqtdci  no        600K  Wed Jan 13 21:06 2021  pre_pacman_-S_grub
pac-infvbv  no        568K  Wed Jan 13 21:04 2021  post_pacman_-S_openssh
pac-wi9mfu  no        616K  Wed Jan 13 21:04 2021  pre_pacman_-S_openssh
pac-d3gyeb  no        696K  Wed Jan 13 20:56 2021  post_pacman_-S_vim
pac-vfz0xz  no        680K  Wed Jan 13 20:56 2021  pre_pacman_-S_vim
default     yes      1012M  Wed Jan 13 20:11 2021  -
```

After booting into a boot environment:
```
[root@localhost ~]# bieaz info
ZFS property for label:    org:bieaz
Current boot environment:  pac-qqtdci
Root dataset container:    /ROOT/
Root dataset full path:    rpool_n9gyiw/ROOT/pac-qqtdci
Boot dataset container:    /BOOT/
Boot dataset full path:    bpool_n9gyiw/BOOT/pac-qqtdci
Root pool name:            rpool_n9gyiw
Boot pool name:            bpool_n9gyiw
[root@localhost ~]# findmnt /
TARGET SOURCE                       FSTYPE OPTIONS
/      rpool_n9gyiw/ROOT/pac-qqtdci zfs    rw,relatime,xattr,posixacl
[root@localhost ~]# findmnt /boot
TARGET SOURCE                       FSTYPE OPTIONS
/boot  bpool_n9gyiw/BOOT/pac-qqtdci zfs    rw,relatime,xattr,posixacl
```

`zfs mount`:
```
[root@localhost ~]# zfs mount
rpool_n9gyiw/ROOT/pac-qqtdci    /
rpool_n9gyiw/DATA/default/opt   /opt
rpool_n9gyiw/DATA/default/srv   /srv
rpool_n9gyiw/DATA/default/home  /home
rpool_n9gyiw/DATA/default/var/lib/lxc  /var/lib/lxc
rpool_n9gyiw/DATA/default/var/lib/AccountsService  /var/lib/AccountsService
bpool_n9gyiw/BOOT/pac-qqtdci    /boot
rpool_n9gyiw/DATA/default/var/www  /var/www
rpool_n9gyiw/DATA/default/var/log  /var/log
rpool_n9gyiw/DATA/default/root  /root
rpool_n9gyiw/DATA/default/var/spool  /var/spool
rpool_n9gyiw/DATA/default/var/games  /var/games
rpool_n9gyiw/DATA/default/var/lib/nfs  /var/lib/nfs
rpool_n9gyiw/DATA/default/usr/local  /usr/local
rpool_n9gyiw/DATA/default/var/lib/docker  /var/lib/docker
rpool_n9gyiw/DATA/default/var/tmp  /var/tmp
```

Create boot environment with pacman hook:
```
[root@localhost ~]# pacman -S vi
...
:: Running pre-transaction hooks...
(1/1) Performing ZFS pre snapshots for the following configurations...
==> Boot environment: proceed, ROZB3_PAC_SKIP=1 not set
...
==> Boot environment: created pre_pacman_-S_vi
...
:: Running post-transaction hooks...
(1/2) Arming ConditionNeedsUpdate...
(2/2) Performing ZFS post snapshots for the following configurations...
==> Boot environment: proceed, ROZB3_PAC_SKIP=1 not set
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-linux
Found initrd image: /boot/initramfs-linux.img
Found fallback initrd image(s) in /boot: initramfs-linux-fallback.img
Detecting boot environments ...
Found linux image: /tmp/bieaz-bemnt/pac-d4rqr5/boot/vmlinuz-
...
Added 8 ZFS boot environments to GRUB
done
GRUB menu updated
Created successfully
==> Boot environment: created post_pacman_-S_vi
```
Dataset layout after creating some boot environments:
```
[root@localhost ~]# zfs list
NAME                                                USED  AVAIL     REFER  MOUNTPOINT
bpool_n9gyiw                                        109M  3.52G       96K  /boot
bpool_n9gyiw/BOOT                                   108M  3.52G       96K  none
bpool_n9gyiw/BOOT/default                           108M  3.52G      108M  legacy
bpool_n9gyiw/BOOT/pac-0znryh                          8K  3.52G      108M  legacy
bpool_n9gyiw/BOOT/pac-d3gyeb                          8K  3.52G      108M  legacy
bpool_n9gyiw/BOOT/pac-d4rqr5                          0B  3.52G      108M  legacy
bpool_n9gyiw/BOOT/pac-infvbv                          8K  3.52G      108M  legacy
bpool_n9gyiw/BOOT/pac-qqtdci                         92K  3.52G      108M  legacy
bpool_n9gyiw/BOOT/pac-vfz0xz                          8K  3.52G      108M  legacy
bpool_n9gyiw/BOOT/pac-vnj3yc                          0B  3.52G      108M  legacy
bpool_n9gyiw/BOOT/pac-wi9mfu                          8K  3.52G      108M  legacy
rpool_n9gyiw                                       1.00G  5.29G      192K  /
rpool_n9gyiw/DATA                                  4.79M  5.29G      192K  none
rpool_n9gyiw/DATA/default                          4.60M  5.29G      192K  /
rpool_n9gyiw/DATA/default/home                      192K  5.29G      192K  /home
rpool_n9gyiw/DATA/default/opt                       192K  5.29G      192K  /opt
rpool_n9gyiw/DATA/default/root                      216K  5.29G      216K  /root
rpool_n9gyiw/DATA/default/srv                       208K  5.29G      208K  /srv
rpool_n9gyiw/DATA/default/usr                       456K  5.29G      192K  /usr
rpool_n9gyiw/DATA/default/usr/local                 264K  5.29G      264K  /usr/local
rpool_n9gyiw/DATA/default/var                      3.18M  5.29G      192K  /var
rpool_n9gyiw/DATA/default/var/games                 192K  5.29G      192K  /var/games
rpool_n9gyiw/DATA/default/var/lib                   960K  5.29G      192K  /var/lib
rpool_n9gyiw/DATA/default/var/lib/AccountsService   192K  5.29G      192K  /var/lib/AccountsService
rpool_n9gyiw/DATA/default/var/lib/docker            192K  5.29G      192K  /var/lib/docker
rpool_n9gyiw/DATA/default/var/lib/lxc               192K  5.29G      192K  /var/lib/lxc
rpool_n9gyiw/DATA/default/var/lib/nfs               192K  5.29G      192K  /var/lib/nfs
rpool_n9gyiw/DATA/default/var/log                  1.25M  5.29G     1.25M  /var/log
rpool_n9gyiw/DATA/default/var/spool                 200K  5.29G      200K  /var/spool
rpool_n9gyiw/DATA/default/var/tmp                   240K  5.29G      240K  /var/tmp
rpool_n9gyiw/DATA/default/var/www                   192K  5.29G      192K  /var/www
rpool_n9gyiw/ROOT                                  1019M  5.29G      192K  none
rpool_n9gyiw/ROOT/default                          1012M  5.29G      980M  /
rpool_n9gyiw/ROOT/pac-0znryh                        600K  5.29G      980M  /
rpool_n9gyiw/ROOT/pac-d3gyeb                        696K  5.29G      972M  /
rpool_n9gyiw/ROOT/pac-d4rqr5                        512K  5.29G      980M  /
rpool_n9gyiw/ROOT/pac-infvbv                        568K  5.29G      980M  /
rpool_n9gyiw/ROOT/pac-qqtdci                       2.53M  5.29G      980M  /
rpool_n9gyiw/ROOT/pac-vfz0xz                        680K  5.29G      947M  /
rpool_n9gyiw/ROOT/pac-vnj3yc                        536K  5.29G      980M  /
rpool_n9gyiw/ROOT/pac-wi9mfu                        616K  5.29G      973M  /
```

**Review by rlaager — CHANGES_REQUESTED** [review]
Please remove the "@rlaager" bits, including in the issue URL. If you're intending on maintaining this, you probably want to replace that with your username.

The formatting is messed up on "Prepare the Live Environment" step 4 (and then the second step 4, which would be step 8). But that might just be the GitHub viewer? I haven't tested building this normally.

nit: poolUUID doesn't seem to fit the variable naming convention of the other variables.

In Create Datasets, step 4, you're going to have newlines in the for loop, you don't need semi-colons at the ends of lines.

It's been my practice to wrap long command lines. Granted, that's a matter of taste, but I like to avoid needing scrollbars to see the full command if the browser window is narrower (e.g. half of a 1080p screen vertically is common for me).

Again, a matter of taste, but I've intentionally avoided heredocs (e.g. `tee ... << EOF`). Using a real text editor makes it easier for people to edit things if they make a typo.

I'm not sure why you're setting $INST_ROOT for the root password and then having to echo it to `chpasswd`, rather than just having the user run `passwd` at that point.

Setting `ZPOOL_VDEV_NAME_PATH=1` in everyone's environment is really heavy-handed. Is there a better way to solve that that just applies to GRUB? Also, hasn't GRUB been setting that for a while? I suspect you're confused about something here. `1` is not a useful value. It needs to be either `YES` or `ON` to take effect anyway (which again, GRUB should be setting); see `lib/libzfs/libzfs_pool.c` in the ZFS tree.

Is the EFI mirroring setup that you're doing the normal way that Arch Linux handles this?

In terms of the big picture, why create yet another boot environment script rather than use zsys?

**Comment by ne9z** [comment]
> Please remove the "@rlaager" bits, including in the issue URL.

Replaced with my user name.
> The formatting is messed up

Fixed.
> nit: poolUUID doesn't seem to fit

Replaced with `INST_UUID`. 
>don't need semi-colons

Removed.
> It's been my practice to wrap long command lines.

I understand your point. But how to wrap lines like this one?
```
echo crypt-swap ${DISK}-part4 /dev/urandom swap,cipher=aes-cbc-essiv:sha256,size=256 >> $INST_MNT/etc/crypttab
```
How about this:
```
echo crypt-swap ${DISK}-part4 \
  /dev/urandom \
  swap,cipher=aes-cbc-essiv:sha256,size=256 \
  >> $INST_MNT/etc/crypttab
```
> Using a real text editor

My intention is to automate things a little.

With heredocs, if the value of the variables are correct, there's no possibility of typo when pasting the commands to a terminal.

In contrast, the text editor approach, which is employed by the Debian and Ubuntu guides, the user needs to manually replace the variable, inside the editor, each time after pasting something.  I think this is more error-prone.

For example, in Debian guide:
```
apt install --yes openssh-server

vi /etc/ssh/sshd_config
# Set: PermitRootLogin yes
```

Here I would prefer just one line of `echo PermitRootLogin yes >> /etc/ssh/sshd_config`, as the default value is commented out in the config file.
> I'm not sure why you're setting $INST_ROOT for the root password and then having to echo it to `chpasswd`, rather than just having the user run `passwd` at that point.

Replaced with `passwd`. Same for setting root pool password.

> Setting `ZPOOL_VDEV_NAME_PATH=1` in everyone's environment is really heavy-handed. Is there a better way to solve that that just applies to GRUB? 

But if the user forgot to set this, `grub-mkconfig` will fail when executed. One method is to 
> Also, hasn't GRUB been setting that for a while?

No, with vanilla GRUB this is not fixed.
> It needs to be either `YES` or `ON` to take effect anyway.

My mistake. Changed to `YES`.
> Is the EFI mirroring setup that you're doing the normal way that Arch Linux handles this?

No, I invented that myself.

Previously I was mirroring EFI with `dd`, as described in the Debian/Ubuntu guide.

But it turns out that my server can't boot from a `dd`ed partition. Only manually mount the partition and copy `EFI` folder worked for me.
> In terms of the big picture, why create yet another boot environment script rather than use zsys?

`bieaz` is a simple script ported from FreeBSD [beadm](https://github.com/vermaden/beadm), it only does one job, to manage boot environments on one system.

Contrast `zsys`, which aims to manage ZFS on multiple systems in a data center setting. I think this is unnecessary and a over kill.

`bieaz` works on vanilla GRUB. I digged in the Ubuntu GRUB source a little, and it turns out that they applied lots of Ubuntu-specific patches on GRUB just to make it work with `zsys`.

I don't think Arch Linux, which strives to provide vanilla packages, will accept this kind of approach.

**Comment by ne9z** [comment]
[Ubuntu GRUB source](https://git.launchpad.net/~ubuntu-core-dev/grub/+git/ubuntu/tree/util/grub.d/10_linux.in?h=focal#n133):
```
# We have a more specialized ZFS handler, with multiple system in 10_linux_zfs.
...
```

And their [10_linux_zfs](https://git.launchpad.net/~ubuntu-core-dev/grub/+git/ubuntu/tree/util/grub.d/10_linux_zfs.in?h=focal) does not exist in [vanilla GRUB](https://git.savannah.gnu.org/cgit/grub.git/tree/util/grub.d).

**Comment by ne9z** [comment]
Another point is, `bieaz` supports arbitrary pool names (rpool_UUID) and dataset paths, such as
```
rpool_foo/b/a/r/default
```
This is also supported for boot pool. 

The only requirements for `bieaz` to work are:
1. Root file system dataset nested at least one dataset.
2. Root and boot file system dataset must have the same name.

**Comment by rlaager** [comment]
> Contrast zsys, which aims to manage ZFS on multiple systems in a data center setting.

zsys was written by the _desktop_ team, designed for Ubuntu desktop. It's not a server / data center thing, nor for multiple systems. I'm not sure where you got that impression.

Your concern about vanilla GRUB is understandable.

**Comment by ne9z** [comment]
> zsys was written by the _desktop_ team, designed for Ubuntu desktop. It's not a server / data center thing, nor for multiple systems. I'm not sure where you got that impression.

I admit I haven't really used it, as I don't use Ubuntu, but the following gives me the impression.

From [its readme](https://github.com/ubuntu/zsys):
```
zsysctl machine

Machine management
Synopsis

Machine management

zsysctl machine COMMAND [flags]

Options

  -h, --help   help for machine

Options inherited from parent commands

  -v, --verbose count   issue INFO (-v) and DEBUG (-vv) output

zsysctl machine list

List all the machines and basic information.
Synopsis

List all the machines and basic information.
```

**Comment by rlaager** [comment]
> I admit I haven't really used it, as I don't use Ubuntu, but the following gives me the impression.
...
Yeah, I'm not sure why they use the word "machine" all over the place. I can see how that was confusing.

**Comment by ne9z** [comment]
@rlaager: Any update on this? I've addressed most issues in the above post.

**Review by rlaager — APPROVED** [review]
I'm going to change my review from "Request changes" to "Approve", to stop it being in the "Request changes" state. This is more of a "not Reject" or "soft Approve".

I pulled this locally and rendered it through the usual process. Most, if not all, of the remaining formatting concerns I had are fine locally. So it was just GitHub's .rst rendering being different.

On the one hand, I don't see any huge issues here. On the other hand, I'm not in a great position to personally determine the quality of this from an Arch Linux perspective.

You also submitted a HOWTO for Void Linux (a distro I'd literally never heard of before). Someone (and I don't know their relationship to the distro nor their experience) immediately said was not representative of a good Void Linux install. This gives me some pause.

I'm not sure whether I should be the one to approve this or not. But for whoever needs to make that decision, here are some questions that might be important:

Do you use one of these (Arch/Artix) distros regularly (either as your "daily driver" workstation or in a server environment), and if so, are you using Root-on-ZFS? How long have you been using Arch/Artix? How long have you been using ZFS? Are you planning on maintaining this HOWTO over the long-term? Do you have any relevant experience (maintaining documentation and/or software) you'd like to mention?

You said you based this on the Debian HOWTO, but you've clearly removed a lot of content. For example, you've lost large bits of explanatory text (e.g. in the encryption section, pool features and options, etc.). Can you comment on that decision? I wrote a lot of those bits, so obviously I think they're a good idea. It's fine if you disagree. I'd just like to understand your thought process a bit more.

A couple of nits:
* "Create root pool partition" should have a trailing colon
* I also maintain a couple of very similar HOWTOs (in my case Debian and Ubuntu). I find it useful to compare the diffs to make sure they are as close to in sync as possible. You might want to use `diff -u "Arch Linux Root on ZFS.rst" "Artix Linux Root on ZFS.rst"` and cleanup some of the spurious e.g. whitespace differences.

**Comment by ne9z** [comment]
@rlaager

I'm using Arch Linux on my laptop and Artix on a Dell R720xd server, both Root-on-ZFS, since early 2020. Previously, I used Root-on-ZFS with Debian, Ubuntu, Proxmox and XigmaNAS since 2016. I have also requested a review by an admin from ArchWiki.

I'm planning to maintain this HOWTO over the long-term. As I'm settled on Arch/Artix, I will certainly perform Root-on-ZFS installation, according to this guide, on other computers as well. This guide will be continuously updated from bug reports and my own experience with such installations.

I'm mostly a user and sysadmin. Some relevant experience might be: [boot environment script](https://gitlab.com/m_zhou/bieaz), [its pacman hook](https://gitlab.com/m_zhou/rozb3-pac) and [a pipeline building Arch Live image with ZFS](https://gitlab.com/m_zhou/archiso).

I certainly appreciate the explanatory text in Debian/Ubuntu guide, will add them back later. I omitted them at first, mostly because they occupies huge page space and are not so relevant for users who just want follow the best practices and paste commands. The nits you mentioned will be addressed as well.

About the Void Linux HOWTO: the two developers commented on the guide are the main developers behind ZFSBootMenu, in their opinion, anything using GRUB instead of ZBM is flawed and inconvenient. Currently, ZBM is only available in Void repo, hence "was not representative of a good Void Linux install". They can not say the same thing for other distros.

**Comment by rlaager** [comment]
This seems fine to me to merge, but ideally I'd like a second opinion. @gmelikov @behlendorf ?

File: docs/Getting Started/Arch Linux/Arch Linux Root on ZFS.rst
Comment (gmelikov) [inline docs/Getting Started/Arch Linux/Arch Linux Root on ZFS.rst:363]: Just to be sure - archzfs already ships 2.0 with zstd? Won't there be a case with zfs 0.8 branch?

**Review by gmelikov — APPROVED** [review]
LGTM for me as a first version of doc, but I didn't check distro-specific commands unfortunately (I don't use Arch).

@ne9z looks like we can use one page for both arch and artix instructions with different info in needed sections, but whatever you like as the maintainer of these pages.

File: docs/Getting Started/Arch Linux/Arch Linux Root on ZFS.rst
Comment (ne9z) [reply docs/Getting Started/Arch Linux/Arch Linux Root on ZFS.rst:363]: [archzfs](https://archzfs.com/archzfs/x86_64/) always ships the latest version of zfs packages for both prebulit and DKMS packages. Currently it's 2.0.1.

**Comment by ne9z** [comment]
@gmelikov I'd keep them separated. The divergence might expand in the future, making a single doc unfeasible.

You can just spin up a VM and test them out, if you have time, that is ;-).

**Pull request merged**
